### PR TITLE
#57_model_tsuno_Pull_Request

### PR DESCRIPTION
**Pull request by tsuno0829**
## 取り組んだトレーニングの概要
　数種類のデータセットに対して、PCAや単回帰を実装した。sklearnライブラリを用いることで、わずか2, 3行で実装することができた。
#57　

## Action list
- sklearnを用いた単回帰の実装
- IrisデータセットをPCAを用いて次元削減・可視化
- sklearnのよく使用するデータセットについて詳しく調べた

## レビューして欲しいポイント
- 単回帰では、データを分布させ、そのデータに対しての単回帰の線形な直線を引いて視覚的におおよそ上手くいっていることを確認した点
- PCAでは、3次元と2次元の軸で表現した場合の図を並べ、比較できるようにした。また、データのクラスに色をつけ、視認性を上げた点

## レビュワーへの質問
　単回帰のときに、サンプルコードの中で、variance score(R2_score)をprintすることが含まれていたのですが、どんな場面・目的で使うのでしょうか？variance score(R2_score) = 1に近付くと精度が高いとかかれているのを見かけたのですが、単回帰においては良い傾き（y=ax+bのa）を学習できているという認識であっているのでしょうか？見当違いなことを言っているかもしれませんが、ご教授お願いします。<br>サンプルコードのURL: https://scikit-learn.org/stable/auto_examples/linear_model/plot_ols.html

## 感想
　単回帰及びPCAをsklearnライブラリで簡単に実装できるスキルが身についたと思う。また、機械学習面においてsklearnライブラリが簡易で便利すぎるため、身に着けることは必須であると再認識した。

## 次に取り組むトレーニング
- 単回帰は、限定的で単純な条件でしか使えないので、重回帰などの表現力が高いモデルを用いた回帰を実装・理解できる勉強をする。
- ワークアウトの検証（ビギナー）に取り組む。

### CODE REVIEW

**Comment by takuro-Ishida** [comment]
トレーニングお疲れ様です.
## レビューしてほしいポイントについて
* 視覚的におおよそ上手くいっていることを確認した点については,目視でみておおよそうまくいってそうだなということは確認しました.
* PCAでは,同じクラスに所属するデータは同じ色で表示されており, PCAで2Dと3Dで可視化できてそうなところを確認しました.

## レビュアーへの質問
> variance scoreをprintすることが含まれていたのですが、どんな場面・目的で使うのでしょうか？

これは, おそらく単回帰の結果の評価に使われる値になります.単回帰で大事なのは手元にあるデータを使って未知のデータを予測することです. データをモデルに与えた時にその結果がそのくらい良かったのか？を知るためのものです.

> variance score = 1に近付くと精度が高いとかかれているのを見かけたのですが、単回帰においては良い傾き（y=ax+bのa）を学習できているという認識であっているのでしょうか？

精度が高ければ単回帰においては良い傾きを学習できているのかということであれば,その認識で基本的に問題ないかと思います
しかし「何の」精度がたかいのか,「良い」傾きは何か？という部分が大事だと個人的には思います.
何の精度をみたくてどのような指標を用いているのか, 単回帰で良い傾きってどんなものなのかという部分はとても重要なところなので考えてみると良いかと思います.(何いってるのか,意味わかんねってなったらまた聞いてください)
sklearnのやつでは, R2スコアという基準を用いてデータのモデルの予測精度を評価しているみたいです.

R2スコアとかの話はこのページが個人的に参考になったのでもしよければ読んでみてください
http://www.renom.jp/ja/notebooks/tutorial/basic_algorithm/correlation_determination/notebook.html

**Review by takuro-Ishida — APPROVED** [review]
石田的には
> 獲得スキルセット
> * irisデータセットに対してPCAをかける
> * Boston house-pricesに対して線形回帰を行い予測できる

単回帰の評価の仕方を調べてもらえればモデルの獲得スキルセットの部分は習得できているのかなと思っているのでApproveしておきます.

**Comment by tsuno0829** [comment]
takuro-Ishidaさん、レビューありがとうございます。
>精度が高ければ単回帰においては良い傾きを学習できているのかということであれば,その認識で基本的に問題ないかと思います
しかし「何の」精度がたかいのか,「良い」傾きは何か？という部分が大事だと個人的には思います.
何の精度をみたくてどのような指標を用いているのか, 単回帰で良い傾きってどんなものなのかという部分はとても重要なところなので考えてみると良いかと思います.(何いってるのか,意味わかんねってなったらまた聞いてください)
sklearnのやつでは, R2スコアという基準を用いてデータのモデルの予測精度を評価しているみたいです.

回帰の評価方法にも多くの種類があることを初めて知りました。sklearnの回帰の評価手法は、以下のURLだけでも4つあることが分かりました。<br>https://pythondatascience.plavox.info/scikit-learn/%E5%9B%9E%E5%B8%B0%E3%83%A2%E3%83%87%E3%83%AB%E3%81%AE%E8%A9%95%E4%BE%A1

そのサイトのR2_scoreの部分には、
>決定係数 (R2, R-squared, coefficient of determination) は、モデルの当てはまりの良さを示す指標で、最も当てはまりの良い場合、1.0 となります (当てはまりの悪い場合、マイナスとなることもあります)。寄与率 (きよりつ) とも呼ばれます。計算式は以下となります。<br>（数式の乗せ方が分からないのであとで編集します....）

とあります。今まで、分類問題をよく自分は扱っていたのでモデルが未知（テスト）データに対して上手くいっている・いないを判断する評価指標として識別率(accuracy)が当たり前にあったので、回帰にも評価指標があって当然であると再認識しました。その上で、R2_scoreは、数式よりテストデータとモデルの予測した値との距離（分散）を示す標準化？した平均二乗誤差を求めているので（数式はまだのせられてないので上のURLからご参照下さい…）、モデルが上手くデータ点を予測できているかを表す、モデルの当てはまり（汎化性）を調べられる指標だと解釈しました。

>しかし「何の」精度がたかいのか,「良い」傾きは何か？という部分が大事だと個人的には思います.

ですので、回帰における絶対の評価指標はないものの、例えば、「R2_score」の精度が高かったので、単回帰は未知データに対して「良い（汎化性が高い）」傾きを持っているのではないか？と判断することが必要だったのではないかと考えました。
　また、R2_score（決定係数）以外にも回帰における評価方法があるのは分かりましたが、MAEやMSE、RMSEなど見比べてもいまいち使い分けが分からない（数式がそっくりに見える）ので、そこはもう少し深堀りして使い分けられるぐらい詳しくなる必要があると感じました。

**Review by ae14watanabe — APPROVED** [review]
sklearnのモデルを呼び出すことができているのでapproveします。

さて回帰の評価指標が色々あってよく分からない、という件についてですが、「検証」の「常人」レベルで「教師ありのモデル選択と評価」がモロその内容なので、そちらのカテゴリでIssuesを出して色々調べたり試してみた結果をミニレポートしてもらうといいかなと。めちゃくちゃ良い勉強になると思いますよ。

**Comment by ae14watanabe** [comment]
> MAEやMSE、RMSEなど見比べてもいまいち使い分けが分からない（数式がそっくりに見える）ので、そこはもう少し深堀りして使い分けられるぐらい詳しくなる必要があると感じました。

これを検証の常人カテゴリとして、メニュー＋αでやるとめちゃくちゃ良いかなと。
これは自分の思いつきですけど、MSEだと良いけど決定係数だと悪い、みたいな人工データ作って考察とかしたら面白そうですね。良い記事になりそうです。

**Comment by ae14watanabe** [comment]
ということでマージしておきました。お疲れ様でした！
